linode/resource_linode_instance.go

```diff
@@ -1160,16 +1160,44 @@ func resourceLinodeInstanceUpdate(d *schema.ResourceData, meta interface{}) erro
 		d.Partial(false)
 	}
 
-	if d.HasChange("type") {
-		if err = changeInstanceType(&client, instance, d.Get("type").(string), d); err != nil {
-			return err
-		}
-		d.Set("type", d.Get("type").(string))
-	}
+	var rebootInstance bool
+	var diskIDLabelMap map[string]int
 
 	tfDisksOld, tfDisksNew := d.GetChange("disk")
 
-	rebootInstance, diskIDLabelMap, err := updateInstanceDisks(client, d, *instance, tfDisksOld, tfDisksNew)
+	tfDisksOldInterface := tfDisksOld.([]interface{})
+	tfDisksNewInterface := tfDisksNew.([]interface{})
+
+	oldDiskSize := 0
+	newDiskSize := 0
+
+	// Get total amount of disk usage before & after
+	for _, disk := range tfDisksOldInterface {
+		oldDiskSize += disk.(map[string]interface{})["size"].(int)
+	}
+
+	for _, disk := range tfDisksNewInterface {
+		newDiskSize += disk.(map[string]interface{})["size"].(int)
+	}
+
+	if newDiskSize > oldDiskSize {
+		if d.HasChange("type") {
+			if err = changeInstanceType(&client, instance, d.Get("type").(string), d); err != nil {
+				return err
+			}
+			d.Set("type", d.Get("type").(string))
+		}
+		rebootInstance, diskIDLabelMap, err = updateInstanceDisks(client, d, *instance, tfDisksOld, tfDisksNew)
+	} else {
+		rebootInstance, diskIDLabelMap, err = updateInstanceDisks(client, d, *instance, tfDisksOld, tfDisksNew)
```

Comment:
The error returned from `updateInstanceDisks` isn't checked here. If the disk update operation fails but the instance type change operation succeeds, this operation will look like it succeeded when in reality it failed.

Reply:
good call. adding this.